posts/running-background-tasks-python.md

```diff
@@ -0,0 +1,185 @@
+
+![health check](./images/running-background-tasks-python/health-check.png)
+
+
+### Getting to Know Starter Application
+
+The starter application uses python based web framework called `Flask`. It will be used for exposing the REST API endpoints. The flask application is served using the `Gunicorn` HTTP server. We use `honcho` to start our web application as it allows us to run multiple processes at the startup. It uses Procfile for specifying the processes that need to be run. Currently, only Gunicorn is started using honcho.
+
+![services cloud](./images/running-background-tasks-python/lib-cloud.png)
+
+`Pipenv` is used for managing the python dependencies and environment. We use `Docker` for containerising and running our application in an isolated environment. Dockerfile lets us specify how our application should be run. Docker compose lets us orchestrate multiple containers (flask web server and Redis server in our case) to run together and be able to communicate with each other.
+
+### Writing the parser
+
+Let's start with writing a simple parser that accepts a Goodreads book page URL. We will be using requests python library for making an HTTP request to get HTML content of the page. BeautifulSoup is a python library that lets us search, manipulate and create structured markup languages such as HTML, XML, etc. This will allow us to retrieve key information like the book title, author, rating, and description.
+
+```python
+# server.py
+
+def parse_book_link_for_meta_data(bookLink):
+  htmlString = requests.get(bookLink).content # Fetch HTML string of the book information page
+  bsTree = BeautifulSoup(htmlString,"html.parser") # Build a searchable tree using fetched HTML
+  # Find the required book attributes in the tree
+  title = bsTree.find("h1", attrs={"id": "bookTitle"}).string
+  author = bsTree.find("a", attrs={"class": "authorName"}).span.string
+  rating = bsTree.find("span", attrs={"itemprop": "ratingValue"}).string
+  description = ''.join(bsTree.find("div", attrs={"id": "description"}).find("span", attrs={"style": "display:none"}).stripped_strings)
+  return dict(title=title.strip() if title else '',author=author.strip() if author else '',rating=float(rating.strip() if rating else 0),description=description)
+```
+
+We can now write a function called `parse_and_persist_book_info` that calls the above parsing function and persists the value to Redis so that it can be retrieved later. This function along with its arguments will be pushed to queue so that the worker process can execute it. Redis is a key-value store where the key should be unique else it may lead to overwriting of a previous value. Here `generate_redis_key_for_book` is a function that generates a unique key for a given book URL.
+
+```python
+# server.py
+import pickle
+
+#........
+#........
+
+# This generates a unique Redis key against a book URL
+generate_redis_key_for_book = lambda bookURL: 'GOODREADS_BOOKS_INFO:' + bookURL
```

Comment:
I would explain in a little more detail what you're doing here if this is a major piece of the post.